Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Repository/ContentGraph.php

```diff
@@ -187,6 +189,39 @@ public function findParentNodeAggregates(
         return $this->mapQueryBuilderToNodeAggregates($queryBuilder);
     }
 
+    public function findAncestorNodeAggregateIds(NodeAggregateId $entryNodeAggregateId): NodeAggregateIds
+    {
+        $queryBuilderInitial = $this->createQueryBuilder()
+            ->select('n.nodeAggregateId, ch.parentnodeanchor')
+            ->from($this->nodeQueryBuilder->tableNames->node(), 'n')
+            ->innerJoin('n', $this->nodeQueryBuilder->tableNames->hierarchyRelation(), 'ch', 'ch.parentnodeanchor = n.relationanchorpoint')
+            ->innerJoin('ch', $this->nodeQueryBuilder->tableNames->node(), 'c', 'c.relationanchorpoint = ch.childnodeanchor')
+            ->where('ch.contentstreamid = :contentStreamId')
+            ->andWhere('c.nodeaggregateid = :entryNodeAggregateId');
+
+        $queryBuilderRecursive = $this->createQueryBuilder()
+            ->select('pn.nodeAggregateId, ph.parentnodeanchor')
+            ->from('ancestry', 'ch')
+            ->innerJoin('ch', $this->nodeQueryBuilder->tableNames->node(), 'pn', 'pn.relationanchorpoint = ch.parentnodeanchor')
+            ->innerJoin('pn', $this->nodeQueryBuilder->tableNames->hierarchyRelation(), 'ph', 'ph.childnodeanchor = pn.relationanchorpoint')
+            ->where('ph.contentstreamid = :contentStreamId');
+
+        $queryBuilderCte = $this->createQueryBuilder()
+            ->select('pn.nodeAggregateId')
+            ->from('ancestry', 'pn')
+            ->setParameter('contentStreamId', $this->contentStreamId->value)
+            ->setParameter('entryNodeAggregateId', $entryNodeAggregateId->value);
```

Comment:
```suggestion
        $queryBuilderInitial = $this->createQueryBuilder()
            ->select('ch.parentnodeanchor')
            ->from($this->nodeQueryBuilder->tableNames->hierarchyRelation(), 'ch')
            ->innerJoin('ch', $this->nodeQueryBuilder->tableNames->node(), 'c', 'c.relationanchorpoint = ch.childnodeanchor')
            ->where('ch.contentstreamid = :contentStreamId')
            ->andWhere('c.nodeaggregateid = :entryNodeAggregateId');

        $queryBuilderRecursive = $this->createQueryBuilder()
            ->select('ph.parentnodeanchor')
            ->from('ancestry', 'ch')
            ->innerJoin('ch', $this->nodeQueryBuilder->tableNames->hierarchyRelation(), 'ph', 'ph.childnodeanchor = ch.parentnodeanchor')
            ->where('ph.contentstreamid = :contentStreamId');

        $queryBuilderCte = $this->createQueryBuilder()
            ->select('n.nodeAggregateId')
            ->from('ancestry', 'a')
            ->innerJoin('a', $this->nodeQueryBuilder->tableNames->node(), 'n', 'n.relationanchorpoint = a.parentnodeanchor')
            ->setParameter('contentStreamId', $this->contentStreamId->value)
            ->setParameter('entryNodeAggregateId', $entryNodeAggregateId->value);
```

The query could be optimized a bit more (at least for deeper trees), if we merge the node table only after we built the hierarchy.